### ISSUE TITLE
darknet.exe only utilizes 25% of RTX 2080 Ti

### ISSUE THREAD

**Issue by rod-hendricks** [AlexeyAB/darknet#4266]
For some reason, darknet.exe only utilizes 25% of RTX 2080 Ti even when built on cudnn_half using cmake-gui. I notice that even when creating the layers  there is some noticeable speed differences already when comparing it with RTX 2070 on ubuntu. 

Also tried building without cudnn_half and the speed is the same which seems to support that somehow the GPU is not fully utilized at all.

Is there some setting that i need to switch on/off to allow full usage of GPU on windows? 

Windows 10 Pro
RTX 2080 Ti
CUDA 10.1
Cudnn 7.6.2 
Nvidia Driver 436.30

**Comment by AlexeyAB** [comment]
1. What command line do you use?
2. What cfg-file do you use?
3. Show screenshot of GPU-Z or nvidia-smi with utilizes 25% of RTX 2080 Ti

**Comment by LukeAI** [comment]
@rod-hendricks so you've compared with RTX 2070 and found 2080Ti gives you inferior FPS? It's not simply that the 2080Ti is doing the same amount of work with less resources?

**Comment by rod-hendricks** [comment]
I used darknet.exe detector demo cfg/coco.data cfg/yolov3.cfg yolov3-608.weights -ext_output vid.mp4. The cfg file and weights are the default yolov3 files. The 2080 just processes about 15-20fps for 1 video as compared to 60++ fps for the 2070 for the same video and command. Below is the nvidia-smi screenshot. 

![image](https://user-images.githubusercontent.com/52641040/68633750-fe84ba80-052d-11ea-916c-d53b2146baee.png)

@LukeAI sorry forgot to mention performance significantly slower for 2080 compared to 2070. So I don't think 2080Ti is close to even being up to par with 2070 in terms of inference processing which suggests something is not right.

**Comment by LukeAI** [comment]
I use a 2080Ti... hope I'm not being short changed on performance

**Comment by AlexeyAB** [comment]
Try to use this command:
`darknet.exe detector demo cfg/coco.data cfg/yolov3.cfg yolov3-608.weights -ext_output -dont_show vid.mp4`

Do you use 608x608 network resolution in cfg-file?
What CUDA, cuDNN and OpenCV versions do you use?
What is your video resolution?

**Comment by rod-hendricks** [comment]
@AlexeyAB 
When using the command you specified, it did improve speed (by about 10fps) and utilization (~36%) but still no where near where it should be. I tried dual boot with ubuntu on the machine and then i get the results that I expect with the very fast speed when running on the ubuntu version. 

cfg = 416x416
CUDA = 10.1
cuDNN = 7.6.4
OpenCV = 3.4.2
vid Resolution = 1920x1080

One difference i notice is that when running on windows, the creation of architecture and loading of weights slows down when it reaches layer 60+. Very noticeable difference when running it in Ubuntu where this is very fast until layer 106.

@LukeAI I think RTX2080Ti is till fast. I got fed up and installed ubuntu to really check and was able to get good speeds. Roughly 8+fps faster in inference compared to RTX2070 and about 40% faster training time also. This issue seems to be coming out with windows + 2080Ti for now as i can't seem to figure out what I'm missing.

**Comment by AlexeyAB** [comment]
@rod-hendricks 
* Did you compile Darknet on Windows by using Cmake or Legacy-way?
* Check that cuDNN is found in Cmake and show screenshot of Cmake-GUI, like this
![image](https://user-images.githubusercontent.com/4096485/68758708-62d37700-061f-11ea-927e-b1a0ee97d26f.png)

**Comment by rod-hendricks** [comment]
@AlexeyAB I compiled using cmake. My cmake-gui is similar to yours in that it only differs with the ZED_CAMERA parameter switched off as I do not have any. Furthermore, when the model architecture is loaded with the weights, it shows compute capability at 750 and CUDNN_HALF=1. I tried switching this off and resulting speed and gpu utilization is the same. So somehow, even when it says CUDNN_HALF=1, it doesn't perform as such for some reason.

**Comment by AlexeyAB** [comment]
@rod-hendricks 
* Try inserting a video card in another PCIe slot.
* What cuDNN version do you use?
* May be somewhere is a bottleneck on slow  CPU or slow HDD
* May be some hardware bug

**Comment by rod-hendricks** [comment]
@AlexeyAB 

We tried multiple cuDNN versions (i.e. 7.6.1, 7.6.2, 7.6.4) but none of them work on windows. We installed ubuntu on the machine (dual boot) and then we get the speed that we expected so I'm not sure if it is a PCIe slot issue or a hardware issue. 

What it feels like it some sort of windows driver issue of some sort for RTX2080Ti. Would you know of anybody who has tried making your repo run on Windows using RTX2080Ti? So far we have tested 2070, and 2060 both show the expected speeds (both even faster than 2080 on windows).

**Comment by hfassold** [comment]
Hi, of course building darknet with CUDNN _enabled_  on Windows works. You have to download the _right_ CUDNN version which _matches_ the CUDA Toolkit you use, that's very important. All 7.X CUDNN versions are binary compatible, so it does not really matter if you use 7.6 or 7.5 (except for performance - later CUDNN versions will be better optimized).

**Comment by rod-hendricks** [comment]
@hfassold Would you be able to share the cuda toolkit and cudnn version you used on your windows with RTX2080Ti along with the fps speed you get when you test out inferencing? I'd like to use it as a benchmark of what I expect to achieve from 2080 on windows. 

I have tried multiple combinations so maybe i haven't hit the sweet spot yet. Also I have tested different GPU devices on windows and all work well so I am really only experiencing this issue on RTX2080Ti.

**Comment by hfassold** [comment]
Toolkit 9.1, cudnn 7.1
I have a quadro RTX 5000

**Comment by rod-hendricks** [comment]
hmmm so i guess this isn't a good comparison then. I need to compare with someone who has worked on RTX2080 on windows. It's the only one that doesn't give me the expected speeds. Other RTX geforce cards I have tried works well on both windows and ubuntu. RTX2080 works well in ubuntu but not windows and I can't figure out why.

**Comment by rod-hendricks** [comment]
@AlexeyAB 
We noticed though that the speed we get from windows is very close to what we get in ubuntu if we build the repo using cmake .. && make && make install. The cmake command however gives me an error when make install is run as it cannot find the libdark.so file to complete installation. But when running the darknet.exe detector demo command at this point shows the same speed we get from windows. 

In windows installation however the dark.vcxproj is part of the solution when building so not sure if this has any relation. The performance though is identical which is signifcantly slower when building using make in ubuntu.

**Comment by AlexeyAB** [comment]
@rod-hendricks 

> The performance though is identical which is signifcantly slower when building using make in ubuntu.

Did you set in Makefile
```
GPU=1
CUDNN=1
CUDNN_HALF=1
OPENCV=1
```

and then do
``
make clean
make
```

**Comment by rod-hendricks** [comment]
@AlexeyAB 
Yes Makefile specifies those 4 options to be enabled. I didn't do make clean though for the cmake build process and just ran cmake && make && make install (speed is slower and make install runs into error). Should it be cmake && make clean && make && make install instead so that the cmake build matches the fast speed I get when i just do make clean && make?

**Comment by AlexeyAB** [comment]
>  install runs into error

* What errors?

* If you use Cmake - then check that CUDNN and CUDA are enabled

**Comment by rod-hendricks** [comment]
@AlexeyAB 
Apologies, the errors were caused by certain things i didn't clean right.

I did do fresh clones of the repo to clean everything up and then did 3 build scenarios.
1. used cmake with ENABLE CUDNN_HALF OFF and got about 30+fps on ./darknet detector demo command
2. used cmake with ENABLE CUDNN_HALF ON and got about 40+fps on ./darknet detector demo command
3. used make with CUDNN_HALF=1  and got about 60+fps on ./darknet detector demo

For these 3 test, I just played with the cudnn_half option and set gpu, cuda, cudnn, and opencv enabled for all. 

Not sure why i always get lower speed performance on the cmake builds.

**Comment by AlexeyAB** [comment]
@rod-hendricks 
Yes, this is strange.

What GPU usage do you get with Makefile and CUDNN_HALF=1 ?

Check that you have only 1 GPU, only 1 CUDA version and only 1 cuDNN version.

Check that CUDA, cuDNN and CUDNN_HALF are enabled

![image](https://user-images.githubusercontent.com/4096485/69719138-ebbed800-1120-11ea-95d6-520a8688be74.png)


Also try to select `all` CUDA arch: 
![image](https://user-images.githubusercontent.com/4096485/69719076-cc27af80-1120-11ea-98c5-60d21c17fe80.png)

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi,

I have a problem, I'm using 2X RTX 2080 TI for training and each GPU are using maximum 40% of the capacity. I have enabled CUDNN_HALF and I use CUDA 10 and CUDNN 7.5.
Is everything OK?

Thanks in advance.
![1](https://user-images.githubusercontent.com/5433346/70845764-b37d0080-1e67-11ea-8905-2c65f6e41000.jpg)

**Comment by AlexeyAB** [comment]
@zpmmehrdad This is normal. Only on high-end GPUs the cuDNN library can achieve ~90% GPU-usage due to large mini_batch_size by using large VRAM size: RTX Titan 24GB, Tesla 32GB or Quadro RTX 8000 48GB.

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi,

Is there any way to get achieve ~90%? Can I train without -map and reduce the submissions from 16 to 8? because when I decrease the submissions  to 8 I got out of memory.

Thanks.

**Comment by AlexeyAB** [comment]
@zpmmehrdad Yes.

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi, 
I train without -map and reduced the submissions t to 8 but I got out of memory error. I compiled with CUDNN_half=0 and 1 but It didn't make much difference. The maximum GPU usage is 40%. Do you have any suggestions?

Thanks

**Comment by AlexeyAB** [comment]
@zpmmehrdad 

Did you compile Darknet with OpenCV?
What CPU do you use?
Do you train yolov3.cfg or yolov3-tiny.cfg?

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi,

I use CPU Core i9 and I have 2 GPU RTX 2080 Ti. My model is Gaussian with random=1 and network size is default.

**Comment by AlexeyAB** [comment]
May be CPU is a bottleneck if you use multi-GPU.
Also with random=1 it can't reach 90% GPU-Utilization.

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi,

What should I do?
Do you have any suggestion? Can I set random=0 and continue the training?

**Comment by AlexeyAB** [comment]
Yes, you can set random=0 and decrease subdivisions. But mAP perhaps will be lower at 1%.

Or you can buy RTX Titan 24GB

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi,

I have changed random to 0 and decrease subdivisions from 16 to 8 but I got Try to set subdivisions=64 in your cfg-file.

**Comment by AlexeyAB** [comment]
Try batch=60 subdivisions=10 random=0

**Comment by zpmmehrdad** [comment]
@AlexeyAB 
Thank you, I changed batch=60 subdivisions=10 random=0 but GPUs uses ~40%. Where is my problem?

**Comment by AlexeyAB** [comment]
Do you use ssd disk?
Attach your cfg-file in zip

**Comment by zpmmehrdad** [comment]
@AlexeyAB 

Yes I do, 


[GV3.txt](https://github.com/AlexeyAB/darknet/files/3964951/GV3.txt)

**Comment by AlexeyAB** [comment]
What command do you use for training?

**Comment by zpmmehrdad** [comment]
@AlexeyAB Yes I use the latest version and the command is:

darknet.exe detector train obj.obj GV3.cfg -map -dont_show -gpus 0,1

**Comment by AlexeyAB** [comment]
What GPU usage if you use this command? 
`darknet.exe detector train obj.obj GV3.cfg -dont_show`

**Comment by zpmmehrdad** [comment]
@AlexeyAB 

I ran this command: 
darknet.exe detector train obj.obj GV3.cfg -dont_show and  the log is:

![2](https://user-images.githubusercontent.com/5433346/70863712-1b167700-1f60-11ea-92c0-ce2b7bfd3c1e.jpg)

**Comment by AlexeyAB** [comment]
It seems that cuDNN just can't fully load your GPU.
You can't do anything about it.

**Comment by zpmmehrdad** [comment]
> It seems that cuDNN just can't fully load your GPU.
> You can't do anything about it.

Can I compile without cuDNN or decrease the cuDNN version?

**Comment by AlexeyAB** [comment]
@zpmmehrdad 

Show screenshot of GPU-Z if you run this command
`darknet.exe detector train obj.obj GV3.cfg -dont_show -gpus 1`

Also what CPU-usage do you get in this case?

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi,

this is my screenshot

![4](https://user-images.githubusercontent.com/5433346/70995512-f815cf80-20e5-11ea-9eb0-c35af861cb4f.jpg)

**Comment by AlexeyAB** [comment]
On RTX 2070:
* I can achieve 70% GPU-usage by using your cfg-file with [Gaussian_yolo] https://github.com/AlexeyAB/darknet/files/3964951/GV3.txt
* and 75% GPU-usage if use [yolo]


* What is the average images size in your training dataset?
* What is the average number of objects per image in your training dataset?
* What GPU-usage can you get if you use `[yolo] iou_loss=mse` instead of `[Gaussian_yolo] iou_loss=giou` ?

**Comment by zpmmehrdad** [comment]
@AlexeyAB Hi,

I'm using RTX 2080 Ti, 
> * What is the average images size in your training dataset?

- Some images are 3840x2160 and 2704x1520 and 1800x1800

> * What is the average number of objects per image in your training dataset?

- ~5-20

> * What GPU-usage can you get if you use `[yolo] iou_loss=mse` instead of `[Gaussian_yolo] iou_loss=giou` ?

- I will test it and give you the result

**Comment by rod-hendricks** [comment]
I tried multi-gpu training on 2 RTX 2080Ti GPUs with AMD 2950X and get GPU utilization at 85+%. (darknet.exe detector train data/coco.data cfg/yolov3.cfg yolov3.weights -map -gpus 0,1)

What confuses me though is that I still get issues when i do inferencing. my gpu utilization is only at 40% while CPU is maxed out at 100% with processing speed of only 39fps. If I load 2 yolos (1 for each gpu) i get 30fps for each yolo instance. 

This machine I haven't tried building on Linux yet and will only be able to do that next year but somehow i still get sub-expectation results from windows.

**Comment by zpmmehrdad** [comment]
@rod-hendricks Hi,

1. Did you use the latest version?
2. What CUDA and cuDNN version do you use?
3. have you tried Gaussian.cfg?

**Comment by rod-hendricks** [comment]
@zpmmehrdad 

1. Did you use the latest version? For the multi-gpu training no. I checked out from a commit done on August.
2. What CUDA and cuDNN version do you use? CUDA 10.1 cuDNN 7.6.5
3. have you tried Gaussian.cfg? No.

**Comment by spaul13** [comment]
I have build the darknet on windows with CUDNN and CUDA libs as per README mentions. 
But while executing the training on NVIDIA GeForce GTX 1080Ti using the command
darknet.exe classifier train data\new_exp_0106.data cfg\myconvnet.cfg -dont_show

Its taking long time to finish each epoch/iteration. Here is the screenshot of the training
image

I have observed from the performance tab also GPU is merely being used only 1% in few occasions.
image

Here is the output of nvidia-smi
image
This also shows GPU utilization 1% (sometimes 0%)

Can anyone plz tell me why the training is taking so low time even though its needed to be done on GPU? Any suggestion will really be a great help for me. Thanks a lot in advance

**Comment by sleeplessai** [comment]
@spaul13 Switch card to CUDA not 3D.

**Comment by cenit** [comment]
how did you build darknet? please update it and also show building logs
